orchestra/migrations/0018_auto_20151007_1451.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+
+def update_projects(apps, schema_editor):
+    Project = apps.get_model('orchestra', 'Project')
+    WorkflowVersion = apps.get_model('orchestra', 'WorkflowVersion')
+
+    for project in Project.objects.all():
+        project.workflow_version = WorkflowVersion.objects.get(
+            slug=project.workflow_slug)
+        project.save()
+
+
+def update_tasks(apps, schema_editor):
+    Task = apps.get_model('orchestra', 'Task')
+    WorkflowStep = apps.get_model('orchestra', 'WorkflowStep')
+
+    for task in Task.objects.all():
+        task.step = WorkflowStep.objects.get(slug=task.step_slug)
+        task.save()
+
+
+def migrate_worker_certs(apps, schema_editor):
+    Certification = apps.get_model('orchestra', 'Certification')
+    WorkerCertification = apps.get_model('orchestra', 'WorkerCertification')
+
+    # Copy unscoped worker certifications to the new scoped objects.
+    unscoped_worker_certs = WorkerCertification.objects.filter(
+        certification__workflow__isnull=True,
+    )
+    for worker_cert in unscoped_worker_certs:
+        scoped_certs = Certification.objects.filter(
+            workflow__isnull=False,
+            slug=worker_cert.certification.slug,
+        )
+
+        # Since unscoped certifications are 'global', create a new worker_cert
+        # for all new scoped certifications
+        for cert in scoped_certs:
+            WorkerCertification.objects.create(
+                certification=cert,
+                worker=worker_cert.worker,
+                task_class=worker_cert.task_class,
+                role=worker_cert.role
+            )
+
+        # Delete worker certs we successfully copied.
+        # Necessary so we can make Certification.workflow non-nullable.
+        if scoped_certs.exists():
+            worker_cert.delete()
```

Comment:
the only keys I'm aware of are to the Worker and to the Certification. Both are ForeignKeys, so there should be no cascades. 

The delete query django issues (against postgres) looks like:

    DELETE FROM "orchestra_workercertification"
    WHERE "orchestra_workercertification"."id" IN 
        (SELECT U0."id" AS Col1
         FROM "orchestra_workercertification" U0 
             INNER JOIN "orchestra_certification" U1 ON ( U0."certification_id" = U1."id" )
         WHERE U1."workflow_id" IS NULL)

which looks right to me. Agreed that this is scary :ghost: